src/Polly.Core/Registry/ResiliencePipelineRegistry.cs

```diff
@@ -232,22 +228,21 @@ public bool TryAddBuilder<TResult>(TKey key, Action<ResiliencePipelineBuilder<TR
     /// </remarks>
     public async ValueTask DisposeAsync()
     {
-        _disposed = true;
-
-        var pipelines = _pipelines.Values.ToList();
-        _pipelines.Clear();
+        if (_disposed)
+        {
+            return;
+        }
 
-        var registries = _genericRegistry.Values.Cast<IAsyncDisposable>().ToList();
-        _genericRegistry.Clear();
+        _disposed = true;
 
-        foreach (var pipeline in pipelines)
+        foreach (var kv in _pipelines)
         {
-            await pipeline.DisposeHelper.ForceDisposeAsync().ConfigureAwait(false);
+            await kv.Value.DisposeHelper.ForceDisposeAsync().ConfigureAwait(false);
         }
 
-        foreach (var disposable in registries)
+        foreach (var kv in _genericRegistry)
         {
-            await disposable.DisposeAsync().ConfigureAwait(false);
+            await ((IAsyncDisposable)kv.Value).DisposeAsync().ConfigureAwait(false);
```

Comment:
```suggestion
            if (kv.Value is IAsyncDisposable asyncDisposable)
            {
                await asyncDisposable.DisposeAsync().ConfigureAwait(false);
            }
```

Reply:
That doesn't match the previous behavior (unconditional cast) and the entires are all `GenericRegistry` instances, so always implement `IAsyncDisposable`.